src/utils/hdfbshot/fbshot.1.man
.TH fbshot 1 "(C) 2002 Dariusz Swiderski, GPL"
.SH NAME
\fBfbshot\fP - takes screenshots from your framebuffer device
.SH SYNOPSIS
.B fbshot [options] filename
.SH DESCRIPTION
.B fbshot
reads your \fB8,15,16,24 and 32-bit\fP framebuffer device and saves a \fBPNG\fP image
file. You can use it for making screenshots of your framebuffer 
application or of your text or (n)curses applications (just
for fun). You can also use it to show somebody (who lives far away) 
your crazy framebuffer boot logo ;-) You may also use your shots in
presentations or anything. I think this program isn't for nothing.
Send \fBfeedback\fP for more suggestions to use it!
.SH OPTIONS
.TP
.B -h
print usage informations
.TP
.B -c console
specifies which \fBconsole\fP (or virtual terminal - like other 
people call this feature) to grab. Sometimes this changing is 
faster then the framebuffer is refreshed (especially when grabbing
framebuffer apps). If so, you should \fBuse -C console\fP ... This
waits some time before grabbing. You can't use \fB-s\fP in this case.
.TP
.B -d device
specifies which framebuffer device you want to read. If 
you do not use this option, it takes the value from the 
\fBFRAMEBUFFER\fP environment variable. If no FRAMEBUFFER 
environment variable is given or if its value is \fBNULL\fP, the
DEFAULT_FB is taken. (see source code) If you got the official 
binary \fB.deb\fP package, this value is \fB/dev/fb0\fP ... Of 
course, if you received a \fB.rpm\fP, it is the same. And if you
take a look into the binary \fBELF\fP file you can still change
this standard.
.TP
.B -i
turns \fBoff\fP interlacing. If you do not want \fBADAM7 interlacing\fP, 
you should use the \fB-i\fP option.
.TP
.B -s seconds
specifies how many seconds to wait before grabbing. \fBfbshot\fP
does not lock console switching, so you can switch it while 
waiting and so you can use this feature if -c doesn't work 
as normal user and can't log in as root (see BUGS).
.SH WHY PNG
Reasons why I chose PNG for output format:

PNG is small, does everything I or you need for a screenshot. PNG is web 
standard. JPEG or GIF are web standard, too, but both JPEG and GIF are limited.
And libjpeg isn't working at my machine :-) and I needn't choose between libgif
or libungif or between some related full-featured graphic libs (libLUG, 
libMagick). But please notice: in some Netscape versions Netscape has some problems
with PNG when calling directly. It displays them \fBsmaller\fP, and with \fBblack 
bars\fP left and right. Within a HTML page there are \fBno\fP problems I know.
.SH BUGS
\fB-c\fP option does not work as non-root. But I do not want to setuid root
the binary, because this is a \fBsecurity hazard\fP. And I do not like such
hazards! Maybe it works with your kernel. Please mail me (with kernel version),
if so. If you are root, it should change to the console you specified two 
times (which is \fBnot\fP a bug).
.SH TODO
Nice, if some coders would help me:
.nf

    * port it to and test it on other color depths!!!
    * make options to specify compression (I think, that's not important)
    * remove bugs and make bugs (but only unintentionly!)
    * send me mail (suggestions, tips, critics)...
.fi
.SH AUTHOR
If you have any comments, patches, questions, critics, tips or anything 
else (well, feedback) to \fBfbshot\fP, send it to 

    \fBDariusz Swiderski <sfires@sfires.net>\fP

.SH GETTING NEW VERSIONS
You may visit my homepage for newer versions of fbshot:

    \fBhttp://www.sfires.net/fbshot/\fP

You should be able to download \fBtar.gz\fPed source.
.SH THANKS TO (from \fBStephan Beyer\fP <fbshot@s-beyer.de> original man page)
Thanks to \fBGlenn Randers-Pehrson\fP <randeg@alum.rpi.edu> who helped
me with some basic libPNG things.
I also have to thank \fBElliot Lee\fP <sopwith@redhat.com> who sent 
me code for converting 8bpp to 24bpp, although this didn't work at my 
machine. I also didn't implement it into the program, but the code contains 
it...
Also thanks to \fBFrank Kaefer\fP <fbshot@sta.to> who I met at #linux.de 
in IRCnet. Caused by him I noticed, that fbshot had a real heavy bug in 
its first two versions.
\fBliff\fP I also met at IRCnet, channel #c. He helped me with some basic 
bit operation stuff. Also thanks to him. 
Thanks, of course, to everyone who downloaded this program, too. If you send 
me mail I'll call you here (if you want). 
So thanks to 
\fBBrad Douglas\fP <brad@neruo.com>, 
\fBEdmund Haworth\fP <oliver.legion@bigfoot.com>, 
\fBElliot Lee\fP <sopwith@redhat.com>, 
\fBFrank Kaefer\fP <fbshot@sta.to> 
\fBGlenn Randers-Pehrson\fP <randeg@alum.rpi.edu>, 
\fB<imaginate@elektromensch.de>\fP, 
and you? mail me if you want to be called here. 
You may tell me Realname and e-mail and what you are thinking of this program, 
send me patches, comments, questions - what you want...
.SH THANKS TO
\fBStephan Beyer\fP <fbshot@s-beyer.de> for the idea and first code.
\fBPaul Mundt\fP <lethal@chaoticdreams.org> for his further work.
.SH COPYLEFT
Copyleft (C) 2002 \fBDariusz Swiderski\fP <sfires@sfires.net>
.P
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.P
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.P
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


src/utils/hdfbshot/fbshot.c
/*
 * fbshot.c -- FrameBuffer Screen Capture Utility
 * (C)opyright 2002 sfires@sfires.net
 *
 * Originally Written by: Stephan Beyer <PH-Linex@gmx.net>
 * Further changes by: Paul Mundt <lethal@chaoticdreams.org>
 * Rewriten and maintained by: Dariusz Swiderski <sfires@sfires.net>
 * 
 * 	This is a simple program that generates a
 * screenshot of the specified framebuffer device and
 * terminal and writes it to a specified file using
 * the PNG format.
 *
 * See ChangeLog for modifications, CREDITS for credits.
 * 
 * Small changes for Reel Multimedia's HDE 
 *   by Georg Acher, BayCom GmbH (acher(at)baycom.de)
 *
 *
 * fbshot is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either Version 2
 * of the License, or (at your option) any later version.
 *
 * fbshot is distributed in the hope that it will be useful, but
 * WITHOUT ANY  WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License with fbshot; if not, please write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 
 * 02111-1307 USA
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <byteswap.h>
#include <sys/types.h>
#include <asm/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <mntent.h>
#include <errno.h>
#include <sys/utsname.h>

#include <sys/vt.h>
#include <png.h>
#include <linux/fb.h> 

#define DEFAULT_FB      "/dev/fb0"
#define PACKAGE 	"fbshot"
#define VERSION 	"0.3"
#define MAINTAINER_NAME "Dariusz Swiderski"
#define MAINTAINER_ADDR "sfires@sfires.net"

static int waitbfg=0; /* wait before grabbing (for -C )... */


/* some conversion macros */
#define RED565(x)    ((((x) >> (11 )) & 0x1f) << 3)
#define GREEN565(x)  ((((x) >> (5 )) & 0x3f) << 2)
#define BLUE565(x)   ((((x) >> (0)) & 0x1f) << 3)

#define ALPHA1555(x) ((((x) >> (15)) & 0x1 ) << 0)
#define RED1555(x)   ((((x) >> (10)) & 0x1f) << 3)
#define GREEN1555(x) ((((x) >> (5 )) & 0x1f) << 3)
#define BLUE1555(x)  ((((x) >> (0 )) & 0x1f) << 3)

struct picture{
  int xres,yres;
  char *buffer;
  struct fb_cmap *colormap;
  char bps,gray;
};

void FatalError(char* err){
  fprintf(stderr,"An error occured: %s %s\nExiting now...\n",err,strerror(errno));
  fflush(stderr);
  exit (1);
}

void Usage(char *binary){
  printf("Usage: %s [-ghi] [-{C|c} vt] [-d dev] [-s n] filename.png\n", binary);
}

void Help(char *binary){
    printf("FBShot - makes screenshots from framebuffer, v%s\n", VERSION);
    printf("\t\tby Dariusz Swiderski <sfires@sfires.net>\n\n");

    Usage(binary);

    printf("\nPossible options:\n");
    printf("\t-C n  \tgrab from console n, for slower framebuffers\n");
    printf("\t-c n  \tgrab from console n\n");
    printf("\t-d dev\tuse framebuffer device dev instead of default\n");
/* not supported as for now
    printf("\t-g    \tsave a grayscaled PNG\n");
 */
    printf("\t-h    \tprint this usage information\n");
    printf("\t-i    \tturns OFF interlacing\n");
    printf("\t-s n  \tsleep n seconds before making screenshot\n");

    printf("\nSend feedback !!!\n");
}

void chvt(int num){
  int fd;
  if(!(fd = open("/dev/console", O_RDWR)))
    FatalError("cannot open /dev/console");
  if (ioctl(fd, VT_ACTIVATE, num))
    FatalError("ioctl VT_ACTIVATE ");
  if (ioctl(fd, VT_WAITACTIVE, num))
    FatalError("ioctl VT_WAITACTIVE");
  close(fd);
  if (waitbfg)
    sleep (3);
}
int sleepval;
int read_fb(char *device, int vt_num, struct picture *pict){
  int fd, vt_old, i,j;
  struct fb_fix_screeninfo fb_fixinfo;
  struct fb_var_screeninfo fb_varinfo;
  struct vt_stat vt_info;

  if (vt_num!=-1){
    if ((fd = open("/dev/console", O_RDONLY)) == -1)
      FatalError("could not open /dev/console");
    if (ioctl(fd, VT_GETSTATE, &vt_info))
      FatalError("ioctl VT_GETSTATE");
    close (fd);
    vt_old=vt_info.v_active;
  }

  if(!(fd=open(device, O_RDONLY)))
    FatalError("Couldn't open framebuffer device");

  if (ioctl(fd, FBIOGET_FSCREENINFO, &fb_fixinfo))
    FatalError("ioctl FBIOGET_FSCREENINFO");

  if (ioctl(fd, FBIOGET_VSCREENINFO, &fb_varinfo))
    FatalError("ioctl FBIOGET_VSCREENINFO");

  pict->xres=fb_varinfo.xres;
  pict->yres=fb_varinfo.yres;
  pict->bps=fb_varinfo.bits_per_pixel;
  pict->gray=fb_varinfo.grayscale;

  if(fb_fixinfo.visual==FB_VISUAL_PSEUDOCOLOR){
    pict->colormap=(struct fb_cmap*)malloc(sizeof(struct fb_cmap));
    pict->colormap->red=(__u16*)malloc(sizeof(__u16)*(1<<pict->bps));
    pict->colormap->green=(__u16*)malloc(sizeof(__u16)*(1<<pict->bps));
    pict->colormap->blue=(__u16*)malloc(sizeof(__u16)*(1<<pict->bps));
    pict->colormap->transp=(__u16*)malloc(sizeof(__u16)*(1<<pict->bps));
    pict->colormap->start=0;
    pict->colormap->len=1<<pict->bps;
    if (ioctl(fd, FBIOGETCMAP, pict->colormap))
      FatalError("ioctl FBIOGETCMAP");
  }
  if (vt_num!=-1)
    chvt(vt_old);
   
  switch(pict->bps){
  case 15:
    i=2;
    break;
  default:
    i=pict->bps>>3;
  }
   
  if(!(pict->buffer=malloc(pict->xres*pict->yres*i)))
    FatalError("couldnt malloc");

  fprintf(stdout, "Framebuffer %s is %i bytes.\n", device,
                    (fb_varinfo.xres * fb_varinfo.yres * i));
  fprintf(stdout, "Grabbing %ix%i ... \n", fb_varinfo.xres, fb_varinfo.yres);

#ifdef DEBUG
/* Output some more information bout actual graphics mode
 */
  fprintf(stdout, "%ix%i [%i,%i] %ibps %igr\n",
  	fb_varinfo.xres_virtual, fb_varinfo.yres_virtual,
  	fb_varinfo.xoffset, fb_varinfo.yoffset,
  	fb_varinfo.bits_per_pixel, fb_varinfo.grayscale); 
  fprintf(stdout, "FIX: card:%s mem:0x%.8X mem_len:%d visual:%i type:%i type_aux:%i line_len:%i accel:%i\n",
  fb_fixinfo.id,fb_fixinfo.smem_start,fb_fixinfo.smem_len,fb_fixinfo.visual,
  fb_fixinfo.type,fb_fixinfo.type_aux,fb_fixinfo.line_length,fb_fixinfo.accel);
#endif

  fflush(stdout);
  if (vt_num!=-1)
    chvt(vt_num);

  j= (read(fd, pict->buffer, ((pict->xres * pict->yres) * i) )!=
  	(pict->xres * pict->yres *i ));
#ifdef DEBUG
  printf("to read:%i read:%i\n",(pict->xres* pict->yres * i), j);	
#endif
  if (vt_num!=-1)
    chvt(vt_old); 

  if(j)
    FatalError("couldn't read the framebuffer");
  else
    fprintf(stdout,"done.\n");
  close (fd);
  return 0;
}
  
void convert8to32(struct picture *pict){
  int i;
  int j=0;
  __u8 c;
  char *out=(char*)malloc(pict->xres*pict->yres*4);
  for (i=0; i<pict->xres*pict->yres; i++)
  {
    c = ((__u8*)(pict->buffer))[i];
    out[j++]=(char)(pict->colormap->red[c]);
    out[j++]=(char)(pict->colormap->green[c]);
    out[j++]=(char)(pict->colormap->blue[c]);
    out[j++]=(char)(pict->colormap->transp[c]);
  }
  free(pict->buffer);
  pict->buffer=out;
}

void convert1555to32(struct picture *pict){
  int i;
  int j=0;
  __u16 t,c;
  char *out=(char*)malloc(pict->xres*pict->yres*4);
  for (i=0; i<pict->xres*pict->yres; i++)
  {
    c = ( (__u16*)(pict->buffer))[i];
    out[j++]=(char)RED1555(c);
    out[j++]=(char)GREEN1555(c);
    out[j++]=(char)BLUE1555(c);
    out[j++]=(char)ALPHA1555(c);
  }
  free(pict->buffer);
  pict->buffer=out;
}

void convert565to24(struct picture *pict){
  int i;
  int j=0;
  __u16 t,c;
  char *out=(char*)malloc(pict->xres*pict->yres*3);
  for (i=0; i<pict->xres*pict->yres; i++)
  {
    c = ( (__u16*)(pict->buffer))[i];
    out[j++]=(char)RED565(c);
    out[j++]=(char)GREEN565(c);
    out[j++]=(char)BLUE565(c);
  }
  free(pict->buffer);
  pict->buffer=out;
}

void convert32(struct picture *pict){
  int i;
  int j=0;
  __u32 c;
  for (i=0; i<pict->xres*pict->yres; i++)
  {
    c = ( (__u32*)(pict->buffer))[i];
    c = (c&0x0000ff00) | ((c&255)<<16)|((c>>16)&255); // remove alpha, swap R/B
    ( (__u32*)(pict->buffer))[i]=c;
  }
}

static int Write_PNG(struct picture * pict, char *filename, int interlace, int gray){
  png_bytep *row_pointers;
  png_structp png_ptr;
  png_infop info_ptr;
  png_text txt_ptr[4];
  struct utsname *host_info;
  char *out;
  int i;
  int bit_depth=0, color_type;   
  FILE *OUTfd = fopen(filename, "wb");
#ifdef DEBUG
    i=open("pict.bin",O_CREAT|O_WRONLY|O_TRUNC);
    fprintf(stdout, "%i\n",write(i,(void*)pict->buffer,(pict->xres)*(pict->yres)));
    perror("dupa");
    fprintf(stdout, "Writing to %s %ix%i %i\n", filename,(pict->xres),(pict->yres),(pict->xres)*(pict->yres));
    fprintf(stdout, "start: %i, size: %i\n", pict->colormap->start,pict->colormap->len);
    fflush(stdout);
    close(i);
#endif

  if (!OUTfd)
    FatalError("couldn't open output file");

  png_ptr = png_create_write_struct(
        	PNG_LIBPNG_VER_STRING, 
        	(png_voidp)NULL, (png_error_ptr)NULL, (png_error_ptr)NULL);
        
  if (!png_ptr)
    FatalError("couldn't create PNG write struct.");

  info_ptr = png_create_info_struct(png_ptr);
  if (!info_ptr){
    png_destroy_write_struct(&png_ptr, (png_infopp)NULL);
    FatalError("couldn't create PNG info struct.");
  }

/*  host_info=(struct utsname*)malloc(sizeof(struct utsname));
  uname(host_info);*/
  txt_ptr[0].key="Name";
  txt_ptr[0].text="FrameBuffer screenshot";
  txt_ptr[0].compression=PNG_TEXT_COMPRESSION_NONE;
  txt_ptr[1].key="Date";
  txt_ptr[1].text="Current Date";
  txt_ptr[1].compression=PNG_TEXT_COMPRESSION_NONE;
  txt_ptr[2].key="Hostname";
  txt_ptr[2].text="snoopy";
  txt_ptr[2].compression=PNG_TEXT_COMPRESSION_NONE;
  txt_ptr[3].key="Program";
  txt_ptr[3].text=PACKAGE" v."VERSION;
  txt_ptr[3].compression=PNG_TEXT_COMPRESSION_NONE;

  png_set_text(png_ptr, info_ptr, txt_ptr, 4);

  png_init_io(png_ptr, OUTfd);
    
  png_set_compression_level(png_ptr, Z_BEST_COMPRESSION);

  row_pointers=(png_bytep*)malloc(sizeof(png_bytep)*pict->yres);

  switch(pict->bps) {
    case 8:
      convert8to32(pict);
      bit_depth=8;
      color_type=PNG_COLOR_TYPE_RGB_ALPHA;
      for (i=0; i<(pict->yres); i++)
        row_pointers[i]=pict->buffer+i*4*(pict->xres);
      png_set_invert_alpha(png_ptr);
      break;
  
    case 15:
      convert1555to32(pict);
      bit_depth=8;
      color_type=PNG_COLOR_TYPE_RGB_ALPHA;
      for (i=0; i<(pict->yres); i++)
        row_pointers[i]=pict->buffer+i*4*(pict->xres);
      png_set_invert_alpha(png_ptr);
      break;

    case 16:
      convert565to24(pict);
      bit_depth=8;
      color_type=PNG_COLOR_TYPE_RGB;
      for (i=0; i<(pict->yres); i++)
        row_pointers[i]=pict->buffer+i*3*(pict->xres);
      png_set_invert_alpha(png_ptr);
      break;

    case 24:
      bit_depth=8;
      color_type=PNG_COLOR_TYPE_RGB;
      for (i=0; i<(pict->yres); i++)
        row_pointers[i]=pict->buffer+i*3*(pict->xres);
      png_set_invert_alpha(png_ptr);
      break;

    case 32:
      convert32(pict);
      bit_depth=8;
      color_type=PNG_COLOR_TYPE_RGB_ALPHA;
      for (i=0; i<(pict->yres); i++)
        row_pointers[i]=pict->buffer+i*4*(pict->xres);
      png_set_invert_alpha(png_ptr);
      break;
    }
    
    if (bit_depth==0){
        fprintf (stderr, "%d bits per pixel are not yet supported! ", pict->bps);
        fprintf (stderr, "But you may write it...\n");
        exit(1);
    }

    png_set_IHDR(png_ptr, info_ptr, pict->xres, pict->yres, 
        bit_depth, color_type, interlace, 
        PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
    
    png_write_info(png_ptr, info_ptr);

    png_write_image(png_ptr, row_pointers);

    png_write_end(png_ptr, info_ptr);
    /* puh, done, now freeing memory... */
    png_destroy_write_struct(&png_ptr, &info_ptr);
   

    free(row_pointers);
    fclose(OUTfd);
    return 0;
 }



static char optstring[] = "hiC:c:d:s:";
static struct option long_options[] = {
        {"slowcon", 1, 0, 'C'},
        {"console", 1, 0, 'c'},
        {"device", 1, 0, 'd'},
        {"help", 0, 0, 'h'},
        {"noint", 0, 0, 'i'},
        {"sleep", 1, 0, 's'},
        {0, 0, 0, 0}
        };
                                                                
int main(int argc, char **argv){
  char *buffer, *device = NULL, *outfile = argv[argc-1];
  int optc, vt_num=-1;
  struct picture pict;
  int interlace=PNG_INTERLACE_ADAM7;
  int gray=0; /* -1 on ; 0 off ; */
  pict.colormap=NULL;
  sleepval=0;
  for(;;){
    int optind = 0;

    if ((optc = getopt_long(argc, argv, optstring, long_options, &optind)) == -1)
      break;

    switch (optc){
    case 'C':
      waitbfg=1;
    case 'c':
      vt_num=atoi(optarg);
      break;
    case 'd':
      device=optarg;
      break;
/* not supported as for now
    case 'g':
      gray=-1;
      break;
 */
    case 'h':
      Help(argv[0]);
      return 1;
      break;
    case 'i':
      interlace=PNG_INTERLACE_NONE;
      break;
    case 's':
      sleepval=atoi(optarg);

      break;
    default:
      break;
    }
  }

  if ((optind==argc) || (1!=argc-optind)){
    Usage(argv[0]);
    return 1;
  }

  if (NULL==device){
    device=getenv("FRAMEBUFFER");
    if (NULL==device){
      device=DEFAULT_FB;
    }
  }

  read_fb(device, vt_num, &pict);

  printf("Writing %s ...",outfile);fflush(stdout);
      
  Write_PNG(&pict, outfile, interlace, gray );

  if(pict.colormap){
    free(pict.colormap->red);
    free(pict.colormap->green);
    free(pict.colormap->blue);
    free(pict.colormap->transp);
    free(pict.colormap);
  }
  free(pict.buffer);
    
  printf ("done.\n");
    
  return 0;
}        
